src/target/riscv/riscv.c

```diff
@@ -3781,8 +3751,12 @@ static int riscv_openocd_step_impl(struct target *target, int current,
 			return ERROR_FAIL;
 	}
 
-	riscv_reg_t trigger_state[RISCV_MAX_HWBPS] = {0};
-	if (disable_triggers(target, trigger_state) != ERROR_OK)
+	if (riscv_enumerate_triggers(target) != ERROR_OK)
+		return ERROR_FAIL;
+
+	RISCV_INFO(r);
+	bool *wps_to_enable = calloc(sizeof(*wps_to_enable), r->trigger_count);
+	if (disable_watchpoints(target, wps_to_enable) != ERROR_OK)
 		return ERROR_FAIL;
```

Comment:
I believe it would be good to also display an error message in case disable_watchpoints() fails:

```suggestion
	if (disable_watchpoints(target, wps_to_enable) != ERROR_OK) {
		LOG_TARGET_ERROR(target, "Failed to temporarily disable watchpoint triggers before single-step.");
		return ERROR_FAIL;
	}
```

Reply:
Addressed.